README.md
# docx

[![Gem Version](https://badge.fury.io/rb/docx.svg)](https://badge.fury.io/rb/docx)
[![Ruby](https://github.com/ruby-docx/docx/workflows/Ruby/badge.svg)](https://github.com/ruby-docx/docx/actions?query=workflow%3ARuby)
[![Coverage Status](https://coveralls.io/repos/github/ruby-docx/docx/badge.svg?branch=master)](https://coveralls.io/github/ruby-docx/docx?branch=master)
[![Gitter](https://badges.gitter.im/ruby-docx/community.svg)](https://gitter.im/ruby-docx/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

A ruby library/gem for interacting with `.docx` files. currently capabilities include reading paragraphs/bookmarks, inserting text at bookmarks, reading tables/rows/columns/cells and saving the document.

## Usage

### Prerequisites

- Ruby 2.5 or later

### Install

Add the following line to your application's Gemfile:

```ruby
gem 'docx'
```

And then execute:

```shell
bundle install
```

Or install it yourself as:

```shell
gem install docx
```

### Reading

``` ruby
require 'docx'

# Create a Docx::Document object for our existing docx file
doc = Docx::Document.open('example.docx')

# Retrieve and display paragraphs
doc.paragraphs.each do |p|
  puts p
end

# Retrieve and display bookmarks, returned as hash with bookmark names as keys and objects as values
doc.bookmarks.each_pair do |bookmark_name, bookmark_object|
  puts bookmark_name
end
```

Don't have a local file but a buffer? Docx handles those to:

```ruby
require 'docx'

# Create a Docx::Document object from a remote file
doc = Docx::Document.open(buffer)

# Everything about reading is the same as shown above
```

### Rendering html
``` ruby
require 'docx'

# Retrieve and display paragraphs as html
doc = Docx::Document.open('example.docx')
doc.paragraphs.each do |p|
  puts p.to_html
end
```

### Reading tables

``` ruby
require 'docx'

# Create a Docx::Document object for our existing docx file
doc = Docx::Document.open('tables.docx')

first_table = doc.tables[0]
puts first_table.row_count
puts first_table.column_count
puts first_table.rows[0].cells[0].text
puts first_table.columns[0].cells[0].text

# Iterate through tables
doc.tables.each do |table|
  table.rows.each do |row| # Row-based iteration
    row.cells.each do |cell|
      puts cell.text
    end
  end

  table.columns.each do |column| # Column-based iteration
    column.cells.each do |cell|
      puts cell.text
    end
  end
end
```

### Writing

``` ruby
require 'docx'

# Create a Docx::Document object for our existing docx file
doc = Docx::Document.open('example.docx')

# Insert a single line of text after one of our bookmarks
doc.bookmarks['example_bookmark'].insert_text_after("Hello world.")

# Insert multiple lines of text at our bookmark
doc.bookmarks['example_bookmark_2'].insert_multiple_lines_after(['Hello', 'World', 'foo'])

# Remove paragraphs
doc.paragraphs.each do |p|
  p.remove! if p.to_s =~ /TODO/
end

# Substitute text, preserving formatting
doc.paragraphs.each do |p|
  p.each_text_run do |tr|
    tr.substitute('_placeholder_', 'replacement value')
  end
end

# Save document to specified path
doc.save('example-edited.docx')
```

### Writing to tables

``` ruby
require 'docx'

# Create a Docx::Document object for our existing docx file
doc = Docx::Document.open('tables.docx')

# Iterate over each table
doc.tables.each do |table|
  last_row = table.rows.last

  # Copy last row and insert a new one before last row
  new_row = last_row.copy
  new_row.insert_before(last_row)

  # Substitute text in each cell of this new row
  new_row.cells.each do |cell|
    cell.paragraphs.each do |paragraph|
      paragraph.each_text_run do |text|
        text.substitute('_placeholder_', 'replacement value')
      end
    end
  end
end

doc.save('tables-edited.docx')
```

### Advanced

``` ruby
require 'docx'

d = Docx::Document.open('example.docx')

# The Nokogiri::XML::Node on which an element is based can be accessed using #node
d.paragraphs.each do |p|
  puts p.node.inspect
end

# The #xpath and #at_xpath methods are delegated to the node from the element, saving a step
p_element = d.paragraphs.first
p_children = p_element.xpath("//child::*") # selects all children
p_child = p_element.at_xpath("//child::*") # selects first child
```

## 扩展使用
### 写入样式

使用方式，类似下面：
```ruby
# 注意使用最新版
gem 'docx', :git => 'https://github.com/rccgroup/docx.git', :tag => 'v0.6.3-rcc'

gem 'docx', :git => 'https://github.com/rccgroup/docx.git', :branch => 'master'
```

支持传入样式：
- insert_text_before(text, formatting = {})
- insert_text_after(text, formatting = {})
- insert_multiple_lines(text, formatting = {})

也可以针对单个 `textrun` 进行设置：
- textrun.set_text(content, formatting = {})

formatting 支持的参数为：
- :italic => boolean
- :bold => boolean
- :underline => boolean
- :font => 'font_name'
- :font_size => font_size
- :color => color as hex string e.g. 'FF0000' for red


```ruby
doc = Docx::Document.open('format_example.docx')

formatting = {
  italic: false,
  underline: true,
  bold: true,
  font: 'Times New Roman',
  font_size: 20,
  color: 'FF0000'
}

doc.bookmarks["price_index_1"].insert_text_after("100", formatting)
doc.bookmarks["price_index_2"].insert_text_after("50", formatting.merge(italic: true, bold: false))
doc.bookmarks["price_index_3"].insert_text_after("999", formatting.merge(font_size: 10, color: 'yellow'))

doc.bookmarks["name_index"].insert_text_before("小明", formatting.merge(font_size: 12, color: 'blue'))

doc.paragraphs.each do |p|
  puts p
end

doc.paragraphs.each do |paragraph|
  paragraph.each_text_run do |tr|
    tr.substitute("_key_1", "小明")
  end
end

doc.bookmarks["batch_index"].insert_multiple_lines(
  ["行业-1", "行业-2", "行业-3"],
  formatting.merge(font_size: 8, underline: false)
)

doc.save('/Users/rcc0016748/Project/phone-cloud/format_example-edited.docx')
```


参考一个好几年前未合的merge request: https://github.com/ruby-docx/docx/pull/26


## Development

### todo

* Calculate element formatting based on values present in element properties as well as properties inherited from parents
* Default formatting of inserted elements to inherited values
* Implement formattable elements.
* Implement styles.
* Easier multi-line text insertion at a single bookmark (inserting paragraph nodes after the one containing the bookmark)


CHANGELOG.md
# Changelog

## 0.6.5-rcc

- insert_multiple_lines_with_formatting 支持传入 texts, 控制插入同一段落有不同样式的文本。
- paragraph.rb 新增 set_multi_texts 方法。

```ruby
base_doc.bookmarks['notes'].insert_multiple_lines_with_formatting([
  {text: '第一句'},
  {text: '第二句', formatting: formatting},
  {text: '第三句'},
  {
    texts: [
      {text: '第一部分', formatting: {italic: true}},
      {text: '第二部分', formatting: {color: 'FF0000'},
      {text: '第三部分', formatting: {underline: true}}
    ]
  }
])
```

## 0.6.4-rcc

- insert_multiple_texts_with_formatting
- insert_multiple_lines_with_formatting

```ruby
# 文本的样式
formatting = {
  italic: false,
  underline: true,
  bold: true,
  font: 'Times New Roman',
  font_size: 20,
  color: 'FF0000'
}

# 段落的样式 -- xml中缩进的格式 <w:ind w:left="420" w:leftChars="0" w:firstLine="411" w:firstLineChars="196" />
paragraph_formatting = { ind: { left: 420, leftChars: 0, firstLine: 411 } }


base_doc.bookmarks['context'].insert_multiple_texts_with_formatting([
  {text: '工程信息订阅了'},
  {text: '2', formatting: formatting},
  {text: '套，价格是'},
  {text: '1299', formatting: formatting},
  {text: '元。'}
])

base_doc.bookmarks['notes'].insert_multiple_lines_with_formatting([
  {text: '第一句'},
  {text: '第二句', formatting: formatting.merge(paragraph_formatting)},
  {text: '第三句'}
])
```

## 0.6.3-rcc

### add feature

支持传入样式：
- insert_text_before(text, formatting = {})
- insert_text_after(text, formatting = {})
- insert_multiple_lines(text, formatting = {})

也可以针对单个 `textrun` 进行设置：
- textrun.set_text(content, formatting = {})

formatting 支持的参数为：
- :italic => boolean
- :bold => boolean
- :underline => boolean
- :font => 'font_name'
- :font_size => font_size
- :color => color as hex string e.g. 'FF0000' for red


## v0.6.2

### Bug fixes

- Fix `Docx::Document#to_s` fails when given file has `document22.xml.rels` [#112](https://github.com/ruby-docx/docx/pull/112), [#106](https://github.com/ruby-docx/docx/pull/106)

## v0.6.1

### Bug fixes

- Use `Zip::File#glob` to match any `document.xml` [#104](https://github.com/ruby-docx/docx/pull/104)

### Chores

- Enable Coverall's coverage report [#102](https://github.com/ruby-docx/docx/pull/102)
- Add table write example to README.md [#99](https://github.com/ruby-docx/docx/pull/99)
- Replace Travis CI build with GitHub Action [#98](https://github.com/ruby-docx/docx/pull/98)
- Add ruby 3.0 to versions for testing on Travis CI [#97](https://github.com/ruby-docx/docx/pull/97)

## v0.6.0

### Enhancements

- Added support for hyperlinks (implemented [#70](https://github.com/ruby-docx/docx/pull/70) again) by ollieh-m and gopeter [#92](https://github.com/ruby-docx/docx/pull/92)

### Chores

- Drop ruby 2.4 from supporeted versions by satoryu [#93](https://github.com/ruby-docx/docx/pull/93)
- Refactoring `spec_helper` by satoryu [#90](https://github.com/ruby-docx/docx/pull/90)
- Starts measuring code coverage with coveralls by satoryu [#88](https://github.com/ruby-docx/docx/pull/88)

## v0.5.0

### Enhancements

- Added opening streams and outputting to a stream [#66](https://github.com/ruby-docx/docx/pull/66)
- Added supports for Office 365 files [#85](https://github.com/ruby-docx/docx/pull/85)

### Bug fixes

- `Docx::Document` handles a docx file without styles.xml [#81](https://github.com/ruby-docx/docx/pull/81)
- Fixes insert text before after were switched [#84](https://github.com/ruby-docx/docx/pull/84)

## v0.4.0

### Enhancements

- Implement substitute method on TextRun class. [#75](https://github.com/ruby-docx/docx/pull/75)

### Improvements

- Updates dependencies. [#72](https://github.com/ruby-docx/docx/pull/72), [#77](https://github.com/ruby-docx/docx/pull/77)
- Fix: #paragraphs grabs paragraphs in tables. [#76](https://github.com/ruby-docx/docx/pull/76)
- Updates supported ruby versions. [#78](https://github.com/ruby-docx/docx/pull/78)


lib/docx/elements/bookmark.rb
require 'docx/elements/element'

module Docx
  module Elements
    class Bookmark
      include Element
      attr_accessor :name

      def self.tag
        'bookmarkStart'
      end

      def initialize(node)
        @node = node
        @name = @node['w:name']
      end

      # Insert text before bookmarkStart node
      def insert_text_before(text, formatting = {})
        text_run = get_run_before
        text_run.set_text("#{text}#{text_run.text}", formatting)
      end

      # Insert text after bookmarkStart node
      def insert_text_after(text, formatting = {})
        text_run = get_run_after
        text_run.set_text("#{text_run.text}#{text}", formatting)
      end

      # 【新增方法】一整批插入到书签后面，可以为各部分文字定自己的样式。 Oliver.chen 2022-02-11
      #
      # base_doc.bookmarks['context'].insert_multiple_texts_after([
      #   { text: '工程信息订阅了' },
      #   { text: '2', formatting: formatting },
      #   { text: '套，价格是' },
      #   { text: '1299', formatting: formatting },
      #   { text: '元。' }
      # ])
      #
      # Note: 注意一个 bookmark 应该只插入一次。
      #
      # 具体的逻辑:
      # - 找到书签的位置 A
      # - 将插入数组的第一个元素E1，插入到书签 A 之后.(A <- E1)
      # - 剩余的元素以此追加,(A <- E1 <- E2 <- E3 <- E4...)
      def insert_multiple_texts_with_formatting(text_array = [])
        len = text_array.length
        return if len == 0

        # 第一个元素
        new_r = Containers::TextRun.create_with(self)
        new_r.insert_after(self)

        new_r.set_text(text_array[0][:text], text_array[0][:formatting])
        for i in 1..(len - 1)
          another_new_r = Containers::TextRun.create_with(self)
          another_new_r.set_text(text_array[i][:text], text_array[i][:formatting])
          another_new_r.insert_after(new_r)

          new_r = another_new_r
        end
      end

      # 【新增方法】 一整批插入到书签后面，可以为各个段落定自己的样式。 Oliver.chen 2022-02-11
      #
      # formatting = {
      #   italic: false,
      #   underline: true,
      #   bold: true,
      #   font: 'Times New Roman',
      #   font_size: 20,
      #   color: 'FF0000',
      #   ind: { left: 0, leftChars: 0 } --> 设置断落缩进的
      # }
      #
      # base_doc.bookmarks['notes'].insert_multiple_lines_with_formatting([
      #   {text: '第一句'},
      #   {text: '第二句', formatting: formatting},
      #   {text: '第三句'},
      #   {
      #     texts: [
      #       {text: '第一部分', formatting: {italic: true}},
      #       {text: '第二部分', formatting: {color: 'FF0000'},
      #       {text: '第三部分', formatting: {underline: true}}
      #     ]
      #   }
      # ])
      #
      # Note:
      # 1. 注意一个 bookmark 应该只插入一次。
      # 2. 段落有时需要控制缩进的问题
      def insert_multiple_lines_with_formatting(text_array = [])
        # Hold paragraphs to be inserted into, corresponding to the index of the strings in the text array
        paragraphs = []
        paragraph = self.parent_paragraph
        # Remove text from paragraph
        paragraph.blank!
        paragraphs << paragraph
        for i in 0...(text_array.size - 1)
          # Copy previous paragraph
          new_p = paragraphs[i].copy
          # Insert as sibling of previous paragraph
          new_p.insert_after(paragraphs[i])
          paragraphs << new_p
        end

        # Insert text into corresponding newly created paragraphs
        paragraphs.each_index do |index|
          # 如果有 texts，则以这个为主，忽略 text + formatting 属性。
          # 因为在 texts，会包含更加详细文字与样式。
          # 这样就可以支持同一个段落，不同文字带有不同的样式。
          # {
          #   texts: [
          #     { text: 'A', formatting: formatting },
          #     { text: 'B', formatting: formatting },
          #     { text: 'C', formatting: formatting },
          #   ]
          # }
          texts = text_array[index][:texts]
          if texts
            paragraphs[index].set_multi_texts(texts)
          else
            # 一整段使用相同的样式。
            paragraphs[index].apply_formatting(text_array[index][:formatting])
            paragraphs[index].set_text(text_array[index][:text], text_array[index][:formatting])
          end
        end
      end

      # 支持写入样式
      # insert multiple lines starting with paragraph containing bookmark node.
      def insert_multiple_lines(text_array, formatting = {})
        # Hold paragraphs to be inserted into, corresponding to the index of the strings in the text array
        paragraphs = []
        paragraph = self.parent_paragraph
        # Remove text from paragraph
        paragraph.blank!
        paragraphs << paragraph
        for i in 0...(text_array.size - 1)
          # Copy previous paragraph
          new_p = paragraphs[i].copy
          # Insert as sibling of previous paragraph
          new_p.insert_after(paragraphs[i])
          paragraphs << new_p
        end

        # Insert text into corresponding newly created paragraphs
        paragraphs.each_index do |index|
          paragraphs[index].set_text(text_array[index], formatting)
        end
      end

      # Get text run immediately prior to bookmark node
      def get_run_before
        # at_xpath returns the first match found and preceding-sibling returns siblings in the
        # order they appear in the document not the order as they appear when moving out from
        # the starting node
        if not (r_nodes = @node.xpath("./preceding-sibling::w:r")).empty?
          r_node = r_nodes.last
          Containers::TextRun.new(r_node)
        else
          new_r = Containers::TextRun.create_with(self)
          new_r.insert_before(self)
          new_r
        end
      end

      # Get text run immediately after bookmark node
      def get_run_after
        if (r_node = @node.at_xpath("./following-sibling::w:r"))
          Containers::TextRun.new(r_node)
        else
          new_r = Containers::TextRun.create_with(self)
          new_r.insert_after(self)
          new_r
        end
      end
    end
  end
end

lib/docx/containers/paragraph.rb
require 'docx/containers/text_run'
require 'docx/containers/container'
require 'docx/formatting/paragraph_formatting'

module Docx
  module Elements
    module Containers
      class Paragraph
        include Container
        include Elements::Element
        include ParagraphFormatting

        attr_reader :properties_tag

        def self.tag
          'p'
        end

        # Child elements: pPr, r, fldSimple, hlink, subDoc
        # http://msdn.microsoft.com/en-us/library/office/ee364458(v=office.11).aspx
        def initialize(node, document_properties = {})
          @node = node
          @properties_tag = 'pPr'
          @document_properties = document_properties
          @font_size = @document_properties[:font_size]
        end

        # Set text of paragraph
        def text=(content)
          if text_runs.size == 1
            text_runs.first.text = content
          elsif text_runs.size == 0
            new_r = TextRun.create_within(self)
            new_r.text = content
          else
            text_runs.each {|r| r.node.remove }
            new_r = TextRun.create_within(self)
            new_r.text = content
          end
        end

        #【新增方法】使用 set_text 支持 formatting。Oliver.chen 2022-02-11
        def set_text(content, formatting = {})
          self.text = content
          text_runs.each { |tr| tr.apply_formatting(formatting) }
        end

        #【新增方法】设置不同样式的文字。Oliver.chen 2022-02-11
        #
        # texts = [
        #   { text: 'A', formatting: formatting },
        #   { text: 'B', formatting: formatting },
        #   { text: 'C', formatting: formatting },
        # ]
        # 注意：该方法会清掉原来段落的内容。不是追加内容，而是重写。
        def set_multi_texts(texts = [])
          # 先清掉原有的文字
          text_runs.each {|r| r.node.remove }

          len = texts.size
          return if len == 0

          new_r = TextRun.create_within(self)
          new_r.set_text(texts[0][:text], texts[0][:formatting])

          for i in 1..(len - 1)
            another_new_r = Containers::TextRun.create_with(self)
            another_new_r.set_text(texts[i][:text], texts[i][:formatting])
            another_new_r.insert_after(new_r)

            new_r = another_new_r
          end
        end

        # Return text of paragraph
        def to_s
          text_runs.map(&:text).join('')
        end
        alias_method :text, :to_s

        # Return paragraph as a <p></p> HTML fragment with formatting based on properties.
        def to_html
          html = ''
          text_runs.each do |text_run|
            html << text_run.to_html
          end
          styles = { 'font-size' => "#{font_size}pt" }
          styles['text-align'] = alignment if alignment
          html_tag(:p, content: html, styles: styles)
        end


        # Array of text runs contained within paragraph
        def text_runs
          @node.xpath('w:r|w:hyperlink').map { |r_node| Containers::TextRun.new(r_node, @document_properties) }
        end

        # Iterate over each text run within a paragraph
        def each_text_run
          text_runs.each { |tr| yield(tr) }
        end

        def aligned_left?
          ['left', nil].include?(alignment)
        end

        def aligned_right?
          alignment == 'right'
        end

        def aligned_center?
          alignment == 'center'
        end

        def font_size
          size_tag = @node.xpath('w:pPr//w:sz').first
          size_tag ? size_tag.attributes['val'].value.to_i / 2 : @font_size
        end

        private

        # Returns the alignment if any, or nil if left
        def alignment
          # alignment_tag = @node.xpath('.//w:jc').first
          # alignment_tag ? alignment_tag.attributes['val'].value : nil
          formatting[:alignment]
        end

      end
    end
  end
end


lib/docx/formatting/text_run_formatting.rb
require 'docx/formatting/formatting'

module Docx
  module TextRunFormatting
    include Formatting

    def apply_formatting(formatting)
      formatting ||= {}

      if (formatting[:font])
        font_node = add_property('rFonts')
        font_node["w:ascii"] = formatting[:font]
        font_node["w:hAnsi"] = formatting[:font]
      end

      if (formatting[:font_size])
        font_size_node = add_property('sz')
        font_size_node['w:val'] = formatting[:font_size] * 2 # Font size is stored in half-points
      end

      add_property('i') if formatting[:italic]
      add_property('b') if formatting[:bold]

      # 添加下划线，需要指定 val
      if formatting[:underline]
        underline_node = add_property('u')
        underline_node["w:val"] = 'single'
      end

      if (formatting[:color])
        color_node = add_property('color')
        color_node["w:val"] = formatting[:color]
      end
    end

    def parse_formatting
      formatting = {}
      formatting[:italic] = !node.xpath('.//w:i').empty?
      formatting[:bold] = !node.xpath('.//w:b').empty?
      formatting[:underline] = !node.xpath('.//w:u').empty?

      font_node = node.at_xpath('.//w:rFonts')
      formatting[:font] = font_node ? font_node['w:ascii'] : document_properties[:font]
      formatting[:font_size] = font_size

      color_node = node.at_xpath('.//w:color')
      formatting[:color] = color_node ? color_node['w:val'] : nil

      formatting
    end
    alias_method :formatting, :parse_formatting

    def font_size
      size_tag = @node.at_xpath('.//w:sz')
      size_tag ? size_tag.attributes['val'].value.to_i / 2 : @document_properties[:font_size]
    end

    def self.default_formatting(document_properties)
      {
        italic: false,
        bold: false,
        underline: false,
        font: document_properties[:font],
        font_size: document_properties[:font_size],
        color: nil
      }
    end
  end
end


lib/docx/formatting/paragraph_formatting.rb
require 'docx/formatting/formatting'

module Docx
  module ParagraphFormatting
    include Formatting

    def apply_formatting(formatting)
      formatting ||= {}

      # 设置段落属性 -- 缩进等
      for_paragraph = formatting[:paragraph] || {}
      for_paragraph.each do |prop_name, prop_value|
        prop_node = add_property(prop_name)
        prop_value.each do |sub_name, sub_value|
          prop_node["w:#{sub_name}"] = sub_value
        end
      end

      # 设置单个
      # if (formatting[:ind])
      #   ind_node = add_property('ind')

      #   # ind_node["w:left"] = "0"
      #   # ind_node["w:leftChars"] = "0"
      #   # ind_node["w:firstLine"] = "0"
      #   # ind_node["w:firstLineChars"] = "0"
      #   # <w:ind w:left="420" w:leftChars="0" w:firstLine="411" w:firstLineChars="196" />
      #   formatting[:ind].each do |key, value|
      #     ind_node["w:#{key}"] = value
      #   end
      # end

      if (formatting[:alignment])
        alignment_node = add_property('jc')
        alignment_node['w:val'] = formatting[:alignment]
      end
    end

    def parse_formatting
      formatting = {}

      alignment_node = node.at_xpath('.//w:jc')
      formatting[:alignment] = alignment_node ? alignment_node['w:val'] : nil

      formatting
    end
    alias_method :formatting, :parse_formatting

    def self.default_formatting
      {
        alignment: nil
      }
    end
  end
end
